tests/test_utils.py

```diff
@@ -28,3 +45,134 @@ def raises_type_error():
+        bdb_driver, alice_keypair, bob_keypair):
+    import random
+    from coalaip_bigchaindb.utils import make_transfer_tx, order_transactions
+    create_tx = bdb_driver.transactions.prepare(
+        operation='CREATE',
+        signers=alice_keypair['public_key'])
+    transfer_to_bob_tx = make_transfer_tx(bdb_driver, input_tx=create_tx,
+                                          recipients=bob_keypair['public_key'])
+    transfer_back_to_alice_tx = make_transfer_tx(
+        bdb_driver, input_tx=transfer_to_bob_tx, recipients=alice_keypair['public_key'])
+    transfer_back_to_bob_tx = make_transfer_tx(
+        bdb_driver, input_tx=transfer_back_to_alice_tx, recipients=bob_keypair['public_key'])
+
+    correct_order = [transfer_to_bob_tx, transfer_back_to_alice_tx, transfer_back_to_bob_tx]
+    for _ in range(0, 100):
+        assert correct_order == order_transactions(
+            random.sample(correct_order, len(correct_order)))
+
+
+def test_order_empty_transations():
+    from coalaip_bigchaindb.utils import order_transactions
+    ordered_tx = order_transactions([])
+    assert ordered_tx == []
+
+
+def test_order_transactions_fails_with_multiple_endings(
+        bdb_driver, alice_keypair, bob_keypair, carly_keypair):
+    from coalaip_bigchaindb.utils import make_transfer_tx, order_transactions
+    create_tx = bdb_driver.transactions.prepare(
+        operation='CREATE',
+        signers=alice_keypair['public_key'])
+    transfer_to_bob_tx = make_transfer_tx(bdb_driver, input_tx=create_tx,
+                                          recipients=bob_keypair['public_key'])
+    transfer_to_carly_tx = make_transfer_tx(
+        bdb_driver, input_tx=create_tx, recipients=carly_keypair['public_key'])
+
+    with raises(ValueError):
+        # Transfer to both bob and carly should result in a multiple endings
+        # error
+        order_transactions([create_tx, transfer_to_bob_tx, transfer_to_carly_tx])
+
+
+def test_order_transactions_fails_with_cyclic_tx(
+        bdb_driver, alice_keypair, bob_keypair, carly_keypair):
+    from coalaip_bigchaindb.utils import make_transfer_tx, order_transactions
+    create_tx = bdb_driver.transactions.prepare(
+        operation='CREATE',
+        signers=alice_keypair['public_key'])
+    transfer_to_bob_tx = make_transfer_tx(bdb_driver, input_tx=create_tx,
+                                          recipients=bob_keypair['public_key'])
+    transfer_to_carly_tx = make_transfer_tx(
+        bdb_driver, input_tx=transfer_to_bob_tx, recipients=carly_keypair['public_key'])
+    transfer_to_alice_tx = make_transfer_tx(
+        bdb_driver, input_tx=transfer_to_carly_tx, recipients=alice_keypair['public_key'])
+
+    # Modify transfer to bob tx so that it links back to the transfer to alice.
+    # This should create a bob->carly->alice->bob cycle
+    transfer_to_bob_tx['inputs'][0]['fulfills']['txid'] = transfer_to_alice_tx['id']
+
+    with raises(ValueError):
```

Comment:
Why do we fail on a cyclic transfer?

As far as I understand, there is no reason that we should allow it (and it should be possible from a technical side, no?)